src/rp2_common/pico_atomic/pico_atomic.c

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (c) 2024 Stephen Street (stephen@redrocketcomputing.com).
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "hardware/address_mapped.h"
+#include "hardware/regs/watchdog.h"
+#include "hardware/regs/sio.h"
+#include "hardware/sync.h"
+
+#include "pico/config.h"
+
+#ifndef __optimize
+#define __optimize __attribute__((optimize("-Os")))
```

Comment:
is there a reason for Os by default? note i am merging this now as I want to test it with something else, but may fixup later

Reply:
From my experience, you get better code from -Os on the cortex parts. Less and generally faster code, but our mileage may vary.  I can remove if you wish.  
Sort of a adjacent question is if the atomic code should be forced in to SRAM?

Reply:
good question; in similar bits of code we have a `#define` to pick

No, i think `-Os` is fine, and probably as you say better in many circumstances (our default being -O3 is a historic cmake thing, and i am worried about knock on effects on existing stuff if i change it)

Reply:
Note that this makes the code incompatible with Clang which doesn't support the `optimize` attribute which is GCC specific. A correct way to achieve this without relying on non-portable behavior is to set the optimization level for this specific target through the build system.

Reply:
I tend not to use clang on my projects.  I will remove this momentarily.

Reply:
I updated the branching underlying this PR, but since the PR has been merged, closed and reverted I'm not sure how to github to see the PR updates. @kilograham @petrhosek can you reopen this PR?